Setup deployment environment

tlconnor opened the pull request:
Closes #1

bboe commented on .gitignore:
```diff
@@ -22,3 +22,8 @@
+/node_modules
+/public/webpack
+
+.idea
```
We typically recommend putting editor specific ignores in your global gitignore file rather than the project one. Otherwise every project's gitignore file would be full of all the possible editor ignores.

bboe commented on Gemfile:
```diff
@@ -5,17 +5,12 @@ ruby '2.3.0'
-# Use Uglifier as compressor for JavaScript assets
-gem 'uglifier', '>= 1.3.0'
+
+gem 'pg'
```
We intentionally want to use sqlite in development. While you need not stick with it, the intended purpose is to realize the databases need not be the same -- and that there are separate environments who configuration can be different.

tlconnor replied:
Makes sense.  In this case I'll just stick to PG.

bboe commented on Gemfile:
```diff
@@ -34,27 +35,18 @@ group :development, :test do
 
-group :production do
-  gem 'pg'
-  gem 'rails_12factor'
```
Do we still need this gem for serving static assets on heroku?

tlconnor replied:
I assume you are talking about rails_12factor?  It is no longer required as of rails 5.  Perhaps this should be removed from the ropes template app?

(see Rails 5 notes at https://github.com/heroku/rails_12factor)

bboe replied:
Thanks for pointing that out. I have removed it from the base project:
https://github.com/AppFolioOnboarding/base/commit/cf3fc6321561848fa629dfe3e3bda4fab5636e50

If you see anything else that may not be required, or done differently please let me know so I we can keep the base project up to date.

bboe commented on app/components/styles.scss:
```diff
@@ -0,0 +1,12 @@
+	position: absolute;
+	width: 100%;
+	height: 100%;
+}
```
👾  can you configure your editor to add newlines at the end of the file?

tlconnor replied:
Thanks, now added.  As a side note, it could be useful to add something like scss-lint which would capture these sorts of things. https://github.com/brigade/scss-lint

bboe replied:
Do you want to try to configure it for your project and I'll copy whatever you come up with over to the base project?

bboe commented on config/application.rb:
```diff
@@ -15,16 +15,23 @@ class Application < Rails::Application
                     "img-src 'self' http: https:",
-                    "script-src 'self' 'unsafe-eval'",
-                    "style-src 'self' https://cdn.jsdelivr.net"]
+                    script_src,
```
Is this allowing everything, or preventing everything? If the former, can you whitelist the domains you require? Same with style_src.

tlconnor replied:
The idea here is to allow javascript and css to be served via the webpack dev server.  I've also whitelisted the bootstrap CDN to allow the loading of bootstrap.

I forget why I added "http: https:" but it looks like they were not needed.  They are now removed.

bboe replied:
Can we be more specific to whitelist just the dev server? This seems to allow JS and CSS from any domain, including inline CSS and script tags.

bboe replied:
Do you have a response to the above comment? It seems like we've lost some of the security that was previously set up.

bboe commented on config/database.yml:
```diff
@@ -5,21 +5,22 @@
 #
 default: &default
-  adapter: sqlite3
-  pool: 5
```
I'm curious. Does the `pool: 5` setting not work for the test environment?

tlconnor replied:
I guess it depends on the CI config.  By default the tests would be single threaded so a connection pool shouldn't be necessary.  I could be wrong though.

bboe replied:
I was curious if your removal from the default section was due to it breaking in the test environment.

bboe commented on config/routes.rb:
```diff
@@ -1,3 +1,4 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  get '/', to: 'application#index'
```
Can you use `root` for this?

tlconnor replied:
Quite right, my rails rustiness is showing

bboe commented on config/webpack.config.js:
```diff
@@ -0,0 +1,76 @@
+var path = require('path');
+var webpack = require('webpack');
+var StatsPlugin = require('stats-webpack-plugin');
```
Is this the right naming scheme for a module? It looks to be different than `path` and `webpack`.

tlconnor replied:
I have to admit this code is auto generated so was not written by me.

I would say the reason for the naming is that `path` and `webpack` are objects/modules while `StatsPlugin` is a class that can be instantiated.

bboe replied:
I suppose as long as the style guide says things should be done that way that's fine. It does appear to be inconsistent at first glance.

This is the style guide that I believe we try to follow for our JS code: https://github.com/airbnb/javascript

bboe replied:
I'll ignore this for now, however, I think it would be great if there's going to be a ton of javascript to get eslint (or similar) set up to enforce a javascript style guide.

bboe commented on lib/tasks/test.rake:
```diff
@@ -0,0 +1 @@
+task :test => 'webpack:compile'
```
👾

tlconnor replied:
I have no idea what that emoji means!

The intention here is to run "webpack:compile" before "test".  Do you think there is a better way (I haven't used rake in a couple of years)?

bboe replied:
Sorry, I used it in a earlier comment. I use 👾  (space invader) when I see white-space related issues. In this case the trailing newline is missing. Fortunately the lint tools should catch these things, but I guess rake files aren't checked.

bboe replied:
This newline is still missing from here. 👾

bboe commented on test/controllers/application_controller_test.rb:
```diff
@@ -0,0 +1,7 @@
+  test 'should get index' do
+    get '/'
+    assert_response :success
+    assert_template :index
```
In rails5 `assert_template` is gone. Can you use `assert_select` instead to ensure that the body of the HTTP response contains what you expect?

tlconnor replied:
Thanks, fixed.

bboe commented on test/fixtures/images.yml:
```diff
@@ -0,0 +1,7 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
We try to avoid using fixtures. The files are still needed to ensure that the table is cleared on each test run, but otherwise we try not to add anything to them.

tlconnor replied:
So I did this exactly as expected? (this file was auto generated via `rails g model image`)

bboe replied:
Yeah, everyone runs into the same issue. We prefer a clean table to start with for all our tests and using code in the test to create the state we need. So we remove some things that rails generates for us.

bboe commented on test/models/image_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Is this file needed?

tlconnor replied:
It was auto generated when I created the model.  I've removed it for now and will add it back if needed.

bboe commented:
Added my initial round of comments. This looks awesome. I am really excited to see how well this all integrates.

As we discussed before, it looks like you may have some work to do to get circle ci to test your work. When you're ready for another round of review please PM me on slack.

rmacklin commented on Procfile:
```diff
@@ -1 +1,2 @@
 web: bundle exec puma -C config/puma.rb
+webpack: ./node_modules/.bin/webpack-dev-server --config config/webpack.config.js
```
Hi @tlconnor, @bboe shared this PR with me and let me you'll be doing Ropes differently.

I wanted to mention that this Procfile is used on Heroku, where the webpack dev server is irrelevant. It's probably okay because this line would be ignored (though I'm not positive), but you may want to move it to a development-only Procfile

tlconnor replied:
Yeah I'm pretty sure Heroku will ignore the line unless you assign a dyno to it.  I'll move it to a development only Procfile if you like though as you say its probably not necessary

rmacklin commented on .gitignore:
```diff
@@ -22,3 +22,8 @@
 # Don't check in node_modules
 node_modules
+# Added by webpack-rails
+/node_modules
```
Is this necessary given line 24?

tlconnor replied:
You are right, it was automatically added when i added webpack-rails.  Now fixed.

bboe commented on app/models/image.rb:
```diff
@@ -0,0 +1,2 @@
+class Image < ApplicationRecord
```
I see that you've added this model and a migration, however, I don't see a `can create valid image` test for the model.

bboe commented:
@tlconnor, I went through again and found two little things. It also looks like this branch needs a rebase. While you're at that I noticed that there's a lot added to the "set up deployment environment commit". Could you break that up into [atomic commits](http://www.freshconsulting.com/atomic-commits/)? If you're not familiar with the term, in essence, each commit should introduce a small, but independent change. The introduction of the model (and its test, always with the associated test) for instance is an atomic unit as I can then use that model in the console.

rmacklin commented:
> [atomic commits](http://www.freshconsulting.com/atomic-commits/)

Hadn't seen that resource before, thanks for sharing. I also liked this one that was shared in the comments: https://medium.com/@fagnerbrack/one-commit-one-change-3d10b10cebbf#.7ybjh9ip3 and its sequel: https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1#.w8nxj8ymh

tlconnor commented:
@bboe @rmacklin I have split the commit to be more atomic.  The commit "Switch from asset pipeline to webpack" is still fairly large as I didn't find a sensible way to break it down further.

I've also removed the `Image` model as it is not relevant to this first step (it will instead be introduced in step 2)

bboe commented on Gemfile.lock:
```diff
@@ -135,11 +128,6 @@ GEM
       nokogiri (~> 1.6.0)
     rails-html-sanitizer (1.0.3)
       loofah (~> 2.0)
-    rails_12factor (0.0.3)
```
I know typing is probably a bit of a challenge so I won't force you to fix this. Nevertheless, it looks like this gemfile.lock change belongs with the previous commit.

tlconnor replied:
ha, I was hoping you wouldn't notice this one  :)

And you are right, I should have split the Gemfile.lock changes correctly.

bboe replied:
😎

bboe commented:
I agree it's hard to see where to partition that commit. So let's stick with what's there. I would like to see more precise content security policy settings to whitelist only the webpack server.
